spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RecurrentDecoder.scala

```diff
@@ -0,0 +1,244 @@
+    val featureSizes = outputSize.drop(1)
+    output.resize(Array(batchSize, times) ++ featureSizes)
+    // Clone N modules along the sequence dimension.
+    extend(featureSizes)
+
+    /**
+     * currentInput forms a T() type. It contains two elements, hidden and input.
+     * Each time it will feed the cell with T(hidden, input) (or T(input, hidden) depends on
+     * your hidDim and inputDim), and the cell will give a table output containing two
+     * identical elements T(output, output). One of the elements from the cell output is
+     * the updated hidden. Thus the currentInput will update its hidden element with this output.
+     */
+    var i = 1
+    // init state
+    currentInput(hidDim) = if (initState != null) initState
+     else hidden
+
+    while (i <= times) {
+      if (i == 1) {
+        // input at t(0) is last time step of user input
+        currentInput(inputDim) = input
+        cells(i - 1).forward(currentInput)
+      } else {
+        // input at t(i) is output at t(i-1)
+        cells(i - 1).forward(cells(i - 2).output)
+      }
+      i += 1
+    }
+
+    copy(cells.map(x => x.output.toTable[Tensor[T]](inputDim)),
+        output, 0)
+    output
+  }
+
+  override def accGradParameters(input: Tensor[T], gradOutput: Tensor[T]): Unit = {
+    currentGradOutput(hidDim) = gradHidden
+    /**
+     * Since we clone module along the time dimension, the output of each
+     * iteration have been recorded by the cloned modules. Thus, we can
+     * reuse these outputs during the backward operations by copying the
+     * outputs to _input variable.
+     *
+     * The output of Cell(i-1) should be one of the elements fed to the inputs
+     * of Cell(i)
+     * The first module in the cells array accepts zero hidden parameter.
+     */
+
+    var i = times
+    while (i >= 1) {
+      currentGradOutput(inputDim) = gradOutput.select(timeDim, i)
+
+      if (i > 1) {
+        cells(i - 1).regluarized(false)
+        _input = cells(i - 2).output
+      } else {
+        cells(i - 1).regluarized(true)
+        _input(hidDim) = hidden
+        _input(inputDim) = input
+      }
+
+      cells(i - 1).accGradParameters(_input, currentGradOutput)
+      currentGradOutput(hidDim) = cells(i - 1).gradInput.toTable(hidDim)
+      i -= 1
+    }
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    gradInput = gradInputCell
+    gradInputCell.resizeAs(output)
+    currentGradOutput(hidDim) = gradHidden
+    var i = times
+    while (i >= 1) {
+      currentGradOutput(inputDim) = gradOutput.select(timeDim, i)
+
+      if (i > 1) {
+        _input = cells(i - 2).output
+      } else {
+        _input(hidDim) = hidden
+        _input(inputDim) = input
+      }
+
+      cells(i - 1).updateGradInput(_input, currentGradOutput)
+      currentGradOutput(hidDim) = cells(i - 1).gradInput.toTable(hidDim)
```

Comment:
A major problem here is that the gradOutput of cells(i) relies on two factors: one from cells(i+1).gradInput and the other from gradOutput.select(timeDim, i).
Thus, the gradOutput of cells(i) should be the summation of the above two factors, other wise the gradient might be inaccurate.